### ISSUE TITLE
`xonsh -c` incorrectly sets $XONSHRC and $XONSHRC_DIR to empty lists, breaking any further interactive xonsh subprocesses, as of 0.10

### ISSUE THREAD

**Issue by Qyriad** [xonsh/xonsh#4394]
In xonsh 0.9.27, `xonsh -c sh` would set `$XONSHRC` to the relevant RC files. As of xonsh 0.10, `xonsh -c sh` sets $XONSHRC, but to nothing (which xonsh interprets as an empty list), meaning if any subprocess of `xonsh -c` then itself tries to start an interactive shell, it will fail to read *any* RC files at all.

Though probably not the only one, I have discovered at least one functional and practical consequence of this if `xonsh` is your default shell and you use Gnome under GDM. GDM will use `$SHELL -c` to start the session, setting `$XONSHRC=[]` for the Wayland compositor process, and thus any `xonsh` instances spawned by any subprocesses of the compositor will act as if there is no RC file at all, prompting the user to *create* a xonshrc file despite there already being one!

## xonfig (with RC)

<details>

```
$ xonfig
+------------------+--------------------+
| xonsh            | 0.10.0             |
| Python           | 3.9.6              |
| PLY              | 3.11               |
| have readline    | True               |
| prompt toolkit   | 3.0.8              |
| shell type       | prompt_toolkit     |
| history backend  | json               |
| pygments         | 2.9.0              |
| on posix         | True               |
| on linux         | True               |
| distro           | arch               |
| on wsl           | False              |
| on darwin        | False              |
| on windows       | False              |
| on cygwin        | False              |
| on msys2         | False              |
| is superuser     | False              |
| default encoding | utf-8              |
| xonsh encoding   | utf-8              |
| encoding errors  | surrogateescape    |
| on jupyter       | False              |
| jupyter kernel   | None               |
| xontrib 1        | output_search      |
| xontrib 2        | whole_word_jumping |
+------------------+--------------------+
```

</details>

## Expected Behavior
For `xonsh -c "env | rg XONSHRC"` to either print *nothing* (the variable being unset), or to print the proper RC files as it did until 0.10 (and still does when passed `--login`. That is, `xonsh -c` should either *not* set `$XONSHRC` at all, or should set it to the RC files as it did until 0.10.

## Current Behavior
`xonsh -c "env | rg XONSHRC"` prints `XONSHRC=` — `xonsh -c` sets `$XONSHRC`, but not to any value (making it empty).

## Steps to Reproduce
```sh
sh-5.1$ xonsh -c "env | rg XONSHRC="
XONSHRC=
```
versus
```sh
sh-5.1$ xonsh
>>> env | rg XONSHRC=
XONSHRC=/etc/xonshrc:/home/qyriad/.config/xonsh/rc.xsh:/home/qyriad/.xonshrc
```

In 0.9.27, those both would have outputted the latter.

## For community
⬇️  **Please click the 👍 reaction instead of leaving a `+1` or 👍  comment**

**Comment by gforsyth** [comment]
Thanks for reporting @Qyriad !  `bisect` points to #4256 as the culprit for this change.
I agree that we should _definitely_ not be setting it to an empty list -- I _think_ the correct behavior here is to keep it unset unless it's forced by a specific `--rc`, does that sound right to you?

**Comment by Qyriad** [comment]
Unset for `-c` seems the most sensible option to me! That's not what it was doing *before*hand, but I think it's the most correct and sensible behavior.

**Comment by chronitis** [comment]
Apologies for causing this breakage (author of #4256).

Re-reading the code, we currently determine what the "logical" set of RC files is based on `XONSHRC`, `XONSHRC_DIR`, `--interactive`, `--login`, `--rc`, `--no-rc`, `-c` and `script-file` (see `test_script_startup` in `tests/test_main.py` for the logic I _believe_ should be followed here), and then overwrite `XONSHRC` and `XONSHRC_DIR` according to whatever was actually eligible for loading (whether it not they actually exist or can be successfully loaded).

The simplest fix is that there is a bug where if you call `xonshrc_context(rcfiles=None)`, `XONSHRC` isn't touched, but if you call `xonshrc_context(rcfiles=[])`, it's set to an empty list (with the result seen). We can just fix this exact case not to set the environment variable if the list is empty.

There is also another environment variable `LOADED_RC_FILES` which is set to a list of booleans as to whether rc files have been successfully loaded, which you have to interpret as matching the whichever of `XONSHRC + [sorted(d/*.xsh) for d in XONSHRC_DIR]` actually exist - I think this is pretty much useless, but it's been part of the API for a while so it probably can't be arbitarily discarded.

I _think_ it would be better to:

 * Never alter `XONSHRC` or `XONSHRC_DIR`; they'll only be set if they were set externally to override which RC files should be loaded. This changes the current behavior of showing the result of determining what to load (in most cases).
 * Drop `LOADED_RC_FILES` as it is currently defined.
 * Expose the list of loaded (or failed) RC files and dirs by some other mechanism, which won't recursively affect the startup of nested processes; either:
    + As a display item in `xonfig`
    + Via a singleton object in the python environment, eg `rc_status = {"filename.xsh": was_loaded, ...}`
    + One or more environment variables with lists of files actually loaded (which xonsh will never read at startup), eg by redefining `LOADED_RC_FILES`, or picking a new name. This could be extended with a list of all paths considered, or a list of RC files which failed to load, but I think the usecase for setting such a variable in every session is pretty niche.

I can do a PR for either a minimal fix or one of the wider solutions outlined, depending on what you think the right logic would be.

**Comment by gforsyth** [comment]
Hey @chronitis -- first, no worries.  We break things, then we fix them.

I agree that `LOADED_RC_FILES` serves seemingly no purpose and I'm on board with ripping it out.

I'm leaning towards a wider fix, and I think listing the configuration files/dirs in `xonfig` is reasonable.
The main benefit of the env-vars today, as I see them, is mostly as a reference of which files and locations are being used.

I think we can accomplish the same thing with a `xonfig` listing but I'm also open to other suggestions

**Comment by Qyriad** [comment]
Personally I'd prefer paths to all sourced RC files be programmatically retrievable in *some* way, and an environment variable is a decent way of doing that (so that's as it's not an environment variable that influences xonsh subprocesses), but that's just me.

**Comment by chronitis** [comment]
The version in #4408 makes the loaded RC paths available in `xonfig`, but also as an attribute of `XonshSession` (so you can do `__xonsh__.rc_files` to get a list of what was loaded). Would that work for you?

**Comment by Qyriad** [comment]
Yep, that seems fine to me!

**Comment by arkhan** [comment]
I have the same problem, how can I solve it?

Greetings

**Comment by gforsyth** [comment]
Hey @arkhan -- @chronitis has a pull request open with a potential fix if you want to try that out.

**Comment by gforsyth** [comment]
Otherwise, it'll probably be merged in in the next day or so and then available in the next release.

**Comment by gforsyth** [comment]
@Qyriad -- I've just merged in #4408 if you'd like to try out current `main`

**Comment by Qyriad** [comment]
Seems to work fine now; I'll go ahead and close this :)
